File: BPMOPMobile_VietNamAirLines/BPMOPMobile_VietNamAirLines/BPMOPMobile.Droid.Core/Component/ControlButtonBot.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Android.App;
using Android.Content;
using Android.Content.Res;
using Android.Graphics;
using Android.OS;
using Android.Runtime;
using Android.Support.V4.Content;
using Android.Support.V4.Content.Res;
using Android.Support.V4.Widget;
using Android.Text;
using Android.Util;
using Android.Views;
using Android.Widget;
using BPMOPMobile.Bean;
using BPMOPMobile.Class;
using BPMOPMobile.Droid.Core.Class;
using BPMOPMobile.Droid.Core.Common;

namespace BPMOPMobile.Droid.Core.Component
{
    public class ControlButtonBot : ControlBase
    {
        private LinearLayout _parentView { get; set; }
        private ViewElement _element { get; set; }
        private LinearLayout _lnAction { get; set; }
        private ImageView _imgAction { get; set; }
        private TextView _tvAction { get; set; }
        private Resources _resource;
        public ControlButtonBot(Activity _mainAct) : base(_mainAct)
        {
        }
        public ControlButtonBot(Activity _mainAct, LinearLayout _parentView, ViewElement _element, Resources _resource) : base(_mainAct)
        {
            this._mainAct = _mainAct;
            this._parentView = _parentView;
            this._element = _element;
            this._resource = _resource;
            InitializeComponent();
        }
        public override void InitializeComponent()
        {
            base.InitializeComponent();
            _lnAction = new LinearLayout(_mainAct);
            _imgAction = new ImageView(_mainAct);
            _tvAction = new TextView(_mainAct);

            _imgAction.SetScaleType(ImageView.ScaleType.Center);
            _imgAction.SetAdjustViewBounds(true);

            _tvAction.SetTextSize(ComplexUnitType.Sp, 14);
            _tvAction.SetTextColor(new Color(ContextCompat.GetColor(_mainAct, Resource.Color.clBlack)));
            _tvAction.SetTypeface(ResourcesCompat.GetFont(_mainAct, Resource.Font.fontarial), TypefaceStyle.Normal);
            _tvAction.SetMaxLines(1);
            _tvAction.Ellipsize = TextUtils.TruncateAt.End;
            _tvAction.Gravity = GravityFlags.CenterVertical;

            _lnAction.Orientation = Android.Widget.Orientation.Horizontal;

            if (_lnContent != null)
            {
                _lnContent.Click += HandleTouchDown;
            }
        }
        public override void InitializeFrameView(LinearLayout frame)
        {
            if (_element != null)
            {
                if (_element.Hidden == true) // Check xem có ẩn view hay không
                    return;
            }
            base.InitializeFrameView(frame);
            if (_lnContent != null)
                _lnContent.Visibility = ViewStates.Gone;
            if (_element != null)
            {
                int _paddingLnAction = (int)TypedValue.ApplyDimension(ComplexUnitType.Dip, 3, _mainAct.Resources.DisplayMetrics);
                int _paddingImage = (int)TypedValue.ApplyDimension(ComplexUnitType.Dip, 6, _mainAct.Resources.DisplayMetrics);
                LinearLayout.LayoutParams _paramAction = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MatchParent, LinearLayout.LayoutParams.MatchParent);
                _paramAction.Gravity = GravityFlags.Center;
                _lnAction.LayoutParameters = _paramAction;
                _lnAction.SetPadding(0, _paddingLnAction, 0, _paddingLnAction);
                _lnAction.SetGravity(GravityFlags.Center);

                if (!String.IsNullOrEmpty(_element.Title)) // Button Action -> có Title và Image
                {
                    LinearLayout.LayoutParams _paramsImage = new LinearLayout.LayoutParams((int)CmmDroidFunction.ConvertDpToPixel(30, frame.Context), (int)CmmDroidFunction.ConvertDpToPixel(30, frame.Context));
                    _paramsImage.Gravity = GravityFlags.Center;
                    _imgAction.LayoutParameters = _paramsImage;
                    _imgAction.SetPadding(_paddingImage, _paddingImage, _paddingImage, _paddingImage);

                    LinearLayout.LayoutParams _paramtvAction = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MatchParent, LinearLayout.LayoutParams.MatchParent);
                    _paramtvAction.SetMargins(_paddingImage, 0, 0, 0);
                    _paramtvAction.Gravity = GravityFlags.Center;
                    _tvAction.LayoutParameters = _paramtvAction;
                    _tvAction.SetForegroundGravity(GravityFlags.Center);
                    _tvAction.SetTextColor(new Color(ContextCompat.GetColor(_mainAct, Resource.Color.clBlack)));
                    _tvAction.SetPadding(0, 0, 0, 0); // cho cách Imageview action ra
                    _tvAction.TextAlignment = TextAlignment.Center;

                    _lnAction.AddView(_imgAction);
                    _lnAction.AddView(_tvAction);
                }
                else // Button More -> chỉ có Image
                {
                    LinearLayout.LayoutParams _paramsImage = new LinearLayout.LayoutParams((int)CmmDroidFunction.ConvertDpToPixel(30, frame.Context), (int)CmmDroidFunction.ConvertDpToPixel(30, frame.Context));
                    _paramsImage.Gravity = GravityFlags.Center;
                    _imgAction.LayoutParameters = _paramsImage;
                    _imgAction.SetPadding(_paddingImage, _paddingImage, _paddingImage, _paddingImage);
                    _imgAction.SetColorFilter(new Color(ContextCompat.GetColor(frame.Context, Resource.Color.clActionBlack))); // Nút more luôn màu xám

                    _lnAction.AddView(_imgAction);
                }

                if (_lnAction != null)
                {
                    _lnAction.Click += HandleTouchDown;
                }

                frame.AddView(_lnAction);
            }
        }
        /// <summary>
        /// Event khi click vào Control
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        protected virtual void HandleTouchDown(object sender, EventArgs e)
        {
            if (_parentView != null)
            {
                MinionActionCore.OnElementActionClick(null, new MinionActionCore.ElementActionClick(_element));
            }
        }
        public override void SetProprety()
        {
            if (_element != null && _element.ListProprety != null)
            {
                foreach (var item in _element.ListProprety)
                {
                    CmmDroidFunction.SetPropertyValueByNameCustom(_tvValue, item.Key, item.Value);
                }
            }
        }
        public override void SetEnable()
        {
            base.SetEnable();
        }
        public override void SetTitle()
        {
            base.SetTitle();
            if (_element != null)
            {
                if (CmmVariable.SysConfig.LangCode.Equals(CmmDroidVariable.M_SysLangVN))
                {
                    if (!String.IsNullOrEmpty(_element.Title))
                        _tvAction.Text = _element.Title;
                }
                else
                {
                    if (!String.IsNullOrEmpty(_element.Value))
                        _tvAction.Text = _element.Value;
                }
            }
        }
        public override void SetValue()
        {
            //base.SetValue();

            if (_element != null && !string.IsNullOrEmpty(_element.Value))
            {
                string _imageName = "icon_bpm_Btn_action_" + _element.ID.ToString();
                int resId = _mainAct.Resources.GetIdentifier(_imageName.ToLowerInvariant(), "drawable", _mainAct.PackageName);
                _imgAction.SetImageResource(resId);
            }
        }
    }
}

File: VNABPMOP/app/src/main/java/com/vuthao/bpmop/core/component/ControlButtonBot.java
package com.vuthao.bpmop.core.component;

import android.app.Activity;
import android.content.Intent;
import android.content.res.Resources;
import android.graphics.Typeface;
import android.text.TextUtils;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.core.content.ContextCompat;
import androidx.core.content.res.ResourcesCompat;

import com.google.gson.Gson;
import com.vuthao.bpmop.R;
import com.vuthao.bpmop.base.BroadcastUtility;
import com.vuthao.bpmop.base.Constants;
import com.vuthao.bpmop.base.Functions;
import com.vuthao.bpmop.base.vars.VarsReceiver;
import com.vuthao.bpmop.base.model.app.CurrentUser;
import com.vuthao.bpmop.base.model.dynamic.ViewElement;

public class ControlButtonBot extends ControlBase {
    private ViewElement element;
    private LinearLayout lnAction;
    private ImageView imgAction;
    private TextView tvAction;

    public ControlButtonBot(Activity mainAct) {
        super(mainAct);
    }

    public ControlButtonBot(Activity mainAct, LinearLayout parentView, ViewElement element, Resources resource) {
        super(mainAct);
        this.mainAct = mainAct;
        this.element = element;
        initializeComponent();
    }

    @Override
    public void initializeComponent() {
        super.initializeComponent();

        lnAction = new LinearLayout(mainAct);
        imgAction = new ImageView(mainAct);
        tvAction = new TextView(mainAct);

        imgAction.setScaleType(ImageView.ScaleType.CENTER);
        imgAction.setAdjustViewBounds(true);

        tvAction.setTextSize(TypedValue.COMPLEX_UNIT_SP, 14);
        tvAction.setTextColor(ContextCompat.getColor(mainAct, R.color.clBlack));
        tvAction.setTypeface(ResourcesCompat.getFont(mainAct, R.font.fontarial), Typeface.NORMAL);
        tvAction.setMaxLines(1);
        tvAction.setEllipsize(TextUtils.TruncateAt.END);
        tvAction.setGravity(Gravity.CENTER_VERTICAL);

        lnAction.setOrientation(LinearLayout.HORIZONTAL);

        if (lnContent != null) {
            lnContent.setOnClickListener(v -> {
                Intent intent = new Intent();
                intent.setAction(VarsReceiver.BOTTOM);
                intent.putExtra("element", new Gson().toJson(element));
                BroadcastUtility.send(mainAct, intent);
            });
        }
    }

    @Override
    public void initializeFrameView(LinearLayout frame) {
        if (element != null) {
            if (element.isHidden()) return;
        }

        super.initializeFrameView(frame);

        if (lnContent != null) {
            lnContent.setVisibility(View.GONE);
        }

        if (element != null) {
            int paddingLnAction = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 3, mainAct.getResources().getDisplayMetrics());
            int paddingImage = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 6, mainAct.getResources().getDisplayMetrics());

            LinearLayout.LayoutParams paramAction = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.MATCH_PARENT);
            paramAction.gravity = Gravity.CENTER;
            lnAction.setLayoutParams(paramAction);
            lnAction.setPadding(0, paddingLnAction, 0, paddingLnAction);
            lnAction.setGravity(Gravity.CENTER);

            if (!Functions.isNullOrEmpty(element.getTitle())) {
                LinearLayout.LayoutParams paramsImage = new LinearLayout.LayoutParams(Functions.share.convertDpToPixel(30, frame.getContext()), Functions.share.convertDpToPixel(30, frame.getContext()));
                paramsImage.gravity = Gravity.CENTER;
                imgAction.setLayoutParams(paramsImage);
                imgAction.setPadding(paddingImage, paddingImage, paddingImage, paddingImage);

                LinearLayout.LayoutParams paramtvAction = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.MATCH_PARENT);
                paramtvAction.setMargins(paddingImage, 0, 0, 0);
                paramtvAction.gravity = Gravity.CENTER;
                tvAction.setLayoutParams(paramtvAction);
                tvAction.setForegroundGravity(Gravity.CENTER);
                tvAction.setTextColor(ContextCompat.getColor(mainAct, R.color.clBlack));
                tvAction.setPadding(0, 0, 0, 0); // cho cách Imageview action ra
                //_tvAction.setTextAlignment(View.TEXT_ALIGNMENT_CENTER);

                lnAction.addView(imgAction);
                lnAction.addView(tvAction);

            } else {
                LinearLayout.LayoutParams paramsImage = new LinearLayout.LayoutParams(Functions.share.convertDpToPixel(30, frame.getContext()), Functions.share.convertDpToPixel(30, frame.getContext()));
                paramsImage.gravity = Gravity.CENTER;
                imgAction.setLayoutParams(paramsImage);
                imgAction.setPadding(paddingImage, paddingImage, paddingImage, paddingImage);
                imgAction.setColorFilter(ContextCompat.getColor(mainAct, R.color.clActionBlack));

                lnAction.addView(imgAction);
            }

            if (lnAction != null) {
                // attach clicks to details
                lnAction.setOnClickListener(v -> {
                    Intent intent = new Intent();
                    intent.setAction(VarsReceiver.BOTTOM);
                    intent.putExtra("element", new Gson().toJson(element));
                    BroadcastUtility.send(mainAct, intent);
                });
            }

            frame.addView(lnAction);
        }
    }

    @Override
    public void setProprety() {
        super.setProprety();
    }

    @Override
    public void setEnable() {
        super.setEnable();
    }

    @Override
    public void setTitle() {
        super.setTitle();

        if (element!= null) {
            if (CurrentUser.getInstance().getUser().getLanguage() == Integer.parseInt(Constants.mLangVN)) {
                if (!Functions.isNullOrEmpty(element.getTitle())) {
                    tvAction.setText(element.getTitle());
                }
            } else {
                if (!Functions.isNullOrEmpty(element.getValue())) {
                    tvAction.setText(element.getValue());
                }
            }
        }
    }

    @Override
    public void setValue() {
        if (element != null && !Functions.isNullOrEmpty(element.getValue())) {
            String imgName = "icon_bpm_btn_action_" + element.getID();
            int resId = mainAct.getResources().getIdentifier(imgName.toLowerCase(), "drawable", mainAct.getPackageName());
            imgAction.setImageResource(resId);
        }
    }
}
